subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/observability/BuildScanFacade.kt

```diff
@@ -0,0 +1,169 @@
+import org.gradle.instantexecution.InstantExecutionHost
+import org.gradle.instantexecution.extensions.unsafeLazy
+import org.gradle.internal.scan.config.BuildScanPluginApplied
+import org.gradle.kotlin.dsl.*
+import kotlin.reflect.KProperty
+
+
+internal
+interface BuildScanFacade {
+
+    var enterpriseServer: String?
+    var termsOfServiceUrl: String?
+    var termsOfServiceAgree: String?
+    var server: String?
+    var isAllowUntrustedServer: Boolean
+    var isCaptureTaskInputFiles: Boolean
+    // TODO not accessible // var isPublishAlways: Boolean
+    // TODO not accessible // var isPublishOnFailure: Boolean
+
+    fun tag(tag: String)
+
+    fun value(key: String, value: Any?)
+
+    fun instantExecutionStoreAction() {
+        value("instant-execution:enabled", true)
+        value("instant-execution:action", "store")
+    }
+
+    fun instantExecutionLoadAction() {
+        value("instant-execution:enabled", true)
+        value("instant-execution:action", "load")
+    }
+
+    fun instantExecutionStateSize(stateSize: Long, fingerprintSize: Long) {
+        value("instant-execution:state:size", stateSize)
+        value("instant-execution:fingerprint:size", fingerprintSize)
+    }
+}
+
+
+internal
+class DefaultBuildScanFacade(
+    private val host: InstantExecutionHost,
+    private val buildScanPluginApplied: BuildScanPluginApplied
+) : BuildScanFacade {
+
+    private
+    val gradleEnterprise: Any by unsafeLazy {
+        host.currentBuild.gradle.settings.extensionNamed("gradleEnterprise")
+    }
+
+    private
+    val buildScan: Any by unsafeLazy {
+        gradleEnterprise.withGroovyBuilder { getProperty("buildScan") }
+    }
+
+    override var enterpriseServer: String?
+        by stringGroovyProperty("server", lenient = true) { gradleEnterprise }
+
+    override var termsOfServiceUrl: String?
+        by stringGroovyProperty { buildScan }
+
+    override var termsOfServiceAgree: String?
+        by stringGroovyProperty { buildScan }
+
+    override var server: String?
+        by stringGroovyProperty { buildScan }
+
+    override var isAllowUntrustedServer: Boolean
+        by booleanGroovyProperty("allowUntrustedServer") { buildScan }
+
+    override var isCaptureTaskInputFiles: Boolean
+        by booleanGroovyProperty("captureTaskInputFiles") { buildScan }
+
+    override fun tag(tag: String): Unit = buildScan {
+        "tag"(tag)
+    }
+
+    override fun value(key: String, value: Any?): Unit = buildScan {
+        "value"(key, value.toString())
+    }
+
+    private
+    val isBuildScanEnabled: Boolean
+        get() = buildScanPluginApplied.isBuildScanPluginApplied
+
+    private
+    fun Any.extensionNamed(name: String): Any =
+        (this as ExtensionAware).extensions.getByName(name)
+
+    private
+    fun stringGroovyProperty(name: String? = null, lenient: Boolean = false, owner: () -> Any) =
+        GroovyPropertyDelegate(name, lenient, owner) { it as? String }
```

Comment:
:x: `as?` would silently return null for an `Int`,  for instance, that doesn't seem to the intended behaviour, is it?

Reply:
Good catch, fixed